plugins/uefi/fu-uefi-bootmgr.c

```diff
@@ -0,0 +1,402 @@
+		boot_next = entry;
+		efi_error_clear ();
+		break;
+	}
+	if (rc < 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "failed to find boot variable");
+		return FALSE;
+	}
+
+	/* already exists */
+	if (var_data != NULL) {
+		efi_loadopt_attr_set (loadopt, LOAD_OPTION_ACTIVE);
+		rc = efi_set_variable (*guid, name, var_data,
+				       var_data_size, attr, 0644);
+		if (rc < 0) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_FAILED,
+					     "could not set boot variable active");
+			return FALSE;
+		}
+
+	/* create a new one */
+	} else {
+		g_autofree gchar *boot_next_name = NULL;
+		for (guint32 value = 0; value < 0x10000; value++) {
+			gint div = value / (sizeof(set_entries[0]) * 8);
+			gint mod = value % (sizeof(set_entries[0]) * 8);
+			if (set_entries[div] & (1<<mod))
+				continue;
+			boot_next = value;
+			break;
+		}
+		if (boot_next >= 0x10000) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "no free boot variables!");
+			return FALSE;
+		}
+		boot_next_name = g_strdup_printf ("Boot%04" G_GUINT32_FORMAT "X", boot_next & 0xffff);
+		rc = efi_set_variable (efi_guid_global, boot_next_name, opt, opt_size,
+				       EFI_VARIABLE_NON_VOLATILE |
+				       EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				       EFI_VARIABLE_RUNTIME_ACCESS,
+				       0644);
+		if (rc < 0) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "could not set boot variable");
+			return FALSE;
+		}
+	}
+
+	/* TODO: conditionalize this on the UEFI version? */
+	if(!fu_uefi_bootmgr_add_to_boot_order (boot_next, error))
+		return FALSE;
+
+	/* set the boot next */
+	real_boot16 = boot_next;
+	rc = efi_set_variable (efi_guid_global, "BootNext", (guint8 *)&real_boot16, 2,
+			       EFI_VARIABLE_NON_VOLATILE |
+			       EFI_VARIABLE_BOOTSERVICE_ACCESS |
+			       EFI_VARIABLE_RUNTIME_ACCESS,
+			       0644);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "could not set BootNext(%" G_GUINT16_FORMAT ")",
+			     real_boot16);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+gboolean
+fu_uefi_setup_bootnext (const gchar *esp_mountpoint, GError **error)
+{
+	gboolean use_fwup_path = FALSE;
+	gsize loader_sz = 0;
+	gssize opt_size = 0;
+	gssize sz, dp_size = 0;
+	guint32 attributes = LOAD_OPTION_ACTIVE;
+	g_autofree guint16 *loader_str = NULL;
+	g_autofree gchar *fwup_esp_path = NULL;
+	g_autofree gchar *fwup_fs_basename = NULL;
+	g_autofree gchar *fwup_fs_path = NULL;
+	g_autofree gchar *label = NULL;
+	g_autofree gchar *shim_fs_path = NULL;
+	g_autofree guint8 *dp_buf = NULL;
+	g_autofree guint8 *opt = NULL;
+
+	/* check fwup.efi exists */
+	fwup_fs_path = fu_uefi_bootmgr_get_cmd_path (esp_mountpoint, "fwup");
+	if (fwup_fs_path == NULL) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "could not find fwup on ESP");
+		return FALSE;
```

Comment:
So I think it would make sense at this time to just install the file to the ESP using glib.

Reply:
Does fwupd currently "mirror" the fwupd.efi somewhere or does the distro just install it in the ESP e.g. using rpm? Certainly fixing the problem is much better than reporting the problem.